File: java/01-commands/tests/sourcecode/src/main/java/com/training/app/ComplexMethod/SearchInDictionary.java
package com.training.app.ComplexMethod;

import java.net.UnknownHostException;
import java.util.Collections;
import java.util.Dictionary;

public class SearchInDictionary {
    public boolean execute(Dictionary<String, Integer> dimensions, int valueToFind) {
        for (String key : Collections.list(dimensions.keys())) {
            if (dimensions.get(key) == valueToFind) {
                return verifyIpAddress(key);
            }
        }

        return false;
    }

    private boolean verifyIpAddress(String key) {
        try {
            if (key.split("\\.").length == 4 && 
                (Blacklist.isWhiteList(key) || 
                !Blacklist.isBlacklisted(key))) {
                return true;
            } else if (key.split(":").length == 8) {
                return true;
            }
        } catch (UnknownHostException e) {
            e.printStackTrace();
        }

        return false;
    }

    private boolean verifyIpAddressSimple(String key) {
        if (key.split("\\.").length == 4) {
            return true;
        } else if (key.split(":").length == 8) {
            return true;
        }

        return false;
    }
}

File: cpp/01-commands/tests/ComplexMethod/SearchInDictionary.cpp
#include "pch.h"
#include "Blacklist.h"
#include "SearchInDictionary.h"

bool SearchInDictionary::Execute(const std::map<std::string, int>& dimensions, int valueToFind)
{
    for (const auto& [key, value] : dimensions)
    {
        if (value == valueToFind)
        {
            return VerifyIpAddress(key);
        }
    }

    return false;
}

bool SearchInDictionary::VerifyIpAddress(std::string_view key)
{
    if ((std::count(key.begin(), key.end(), '.') == 3) && (Blacklist::IsWhiteList(key) || !Blacklist::IsBlacklisted(key)))
    {
        return true;
    }
    else if (std::count(key.begin(), key.end(), ':') == 7)
    {
        return true;
    }
     
    return false;
}

bool SearchInDictionary::VerifyIpAddressSimple(std::string_view key)
{
    if (std::count(key.begin(), key.end(), '.') == 3)
    {
        return true;
    }
    else if (std::count(key.begin(), key.end(), ':') == 7)
    {
        return true;
    }

    return false;
}